### PR TITLE
Update head and h-label format for hierarchical label classification

### PR DESCRIPTION
**Pull request by sooahleex**
### Summary

- Related to revamp model for h-cls
- Update head of h-label cls with CBAM head
- Update ground truth format of h-label cls
  - For example, this case has label like [-1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 7].
    This is example of cifar100, which has 100 subclasses and 20 superclasses. 
  image

- Only consider `multi-class` case, so for `multi-label` case, I will cover later.

### How to test

- Update unit test and dataset for it

### Checklist

- [X] I have added unit tests to cover my changes.​
- [X] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/classification/efficientnet.py [hunk 1]
```diff
@@ -432,16 +433,24 @@ def _create_model(self) -> nn.Module:
         model.init_weights()
         return model
 
+    def label_smoothing_loss(self, output, target, num_classes, smoothing=0.1):
```
Comment (sovrasov) [inline src/otx/algo/classification/efficientnet.py]: It looks like torch supports label smoothing out of the box: https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html

Comment (harimkang) [reply src/otx/algo/classification/efficientnet.py]: If it's the same, it would be good to change this, but if it's different, I would move this loss function to the `algo/classification/loss` folder instead of this file.

File: src/otx/algo/classification/classifier/base_classifier.py [hunk 2]
```diff
@@ -81,7 +81,6 @@ def __init__(
         self.explainer = ReciproCAM(
             self._head_forward_fn,
             num_classes=head.num_classes,
-            optimize_gap=True,
         )
```
Comment (harimkang) [inline src/otx/algo/classification/classifier/base_classifier.py]: In my opinion, it seems like a good idea to modify this to accept `optimize_gap` without clearing it. The current change is to change everything that used to be `True` to `False`, and i'm not sure what problems this will cause.

```python
def __init__(
    optimize_gap: bool = True,
):
    self.explainer = ReciproCAM(
        optimize_gap=optimize_gap,
    )
```

File: src/otx/algo/classification/efficientnet_v2.py
Comment (harimkang) [inline src/otx/algo/classification/efficientnet_v2.py:1]: This file has been replaced by `timm_model.py`. please apply changes to that file.

File: src/otx/algo/classification/efficientnet.py [hunk 3]
```diff
@@ -265,9 +265,9 @@ def _build_model(self, head_config: dict) -> nn.Module:
         backbone = OTXEfficientNet(version=self.version, pretrained=self.pretrained)
         return ImageClassifier(
             backbone=backbone,
-            neck=GlobalAveragePooling(dim=2),
-            head=HierarchicalLinearClsHead(
-                in_channels=backbone.num_features,
+            neck=nn.Identity(),
+            head=HierarchicalCBAMClsHead(
+                in_channels=1280,
```
Comment (harimkang) [inline src/otx/algo/classification/efficientnet.py]:
```suggestion
                in_channels=backbone.num_features,
```

**Review by harimkang — COMMENTED** [review]
Looks good, I left a few comments, and could you please write some more unit-tests for the classes added to `hlabel_cls_head.py`?

Also, it would be great if you could summarize and share the results of experiments comparing the old head to the new head. (cifar100)

+ @chuneuny-emily I think we should use cifar100 that sooah used in experiments here for H-label cls large performance tests, what do you think? (There is currently no Large-data case for H-label.)

File: src/otx/algo/classification/efficientnet_v2.py
Comment (sooahleex) [reply src/otx/algo/classification/efficientnet_v2.py:1]: Fix in [0f91b61](https://github.com/openvinotoolkit/training_extensions/pull/3810/commits/0f91b614d179a84d4476db57a761a6b2f11a8a7f)

File: src/otx/algo/classification/efficientnet.py (hunk 3 above)
Comment (sooahleex) [reply src/otx/algo/classification/efficientnet.py]: Fix in [0f91b61](https://github.com/openvinotoolkit/training_extensions/pull/3810/commits/0f91b614d179a84d4476db57a761a6b2f11a8a7f)

File: src/otx/algo/classification/classifier/base_classifier.py (hunk 2 above)
Comment (sooahleex) [reply src/otx/algo/classification/classifier/base_classifier.py]: Fix in [0f91b61](https://github.com/openvinotoolkit/training_extensions/pull/3810/commits/0f91b614d179a84d4476db57a761a6b2f11a8a7f)

File: src/otx/algo/classification/heads/hlabel_cls_head.py [hunk 4]
```diff
@@ -353,3 +353,134 @@ def forward(self, feats: tuple[torch.Tensor] | torch.Tensor) -> torch.Tensor:
...
+        head_idx_to_logits_range: the logit range of each heads
+        num_single_label_classes: the number of single label classes
+        empty_multiclass_head_indices: the index of head that doesn't include any label
+            due to the label removing
+        in_channels (int): Number of channels in the input feature map.
+        num_classes (int): Number of total classes.
+        multiclass_loss (dict | None): Config of multi-class loss.
+        multilabel_loss (dict | None): Config of multi-label loss.
+        thr (float | None): Predictions with scores under the thresholds are considered
+                            as negative. Defaults to 0.5.
```
Comment (harimkang) [inline src/otx/algo/classification/heads/hlabel_cls_head.py]:
```suggestion
        num_multiclass_heads (int): Number of multi-class heads.
        num_multilabel_classes (int): Number of multi-label classes.
        head_idx_to_logits_range (dict[str, tuple[int, int]]): the logit range of each heads
        num_single_label_classes (int): the number of single label classes
        empty_multiclass_head_indices (list[int]): the index of head that doesn't include any label
            due to the label removing
        in_channels (int): Number of channels in the input feature map.
        num_classes (int): Number of total classes.
        multiclass_loss (nn.Module): Config of multi-class loss.
        multilabel_loss (nn.Module | None, optional): Config of multi-label loss.
        thr (float, optional): Predictions with scores under the thresholds are considered
                            as negative. Defaults to 0.5.
        init_cfg (dict | None, optional): Initialize configuration key-values, Defaults to None.
        step_size (int, optional): Step size value for HierarchicalCBAMClsHead, Defaults to 7.
```

File: src/otx/algo/classification/torchvision_model.py
Comment (harimkang) [inline src/otx/algo/classification/torchvision_model.py:1]: Does leaving `optimize_gap` to True work well with torchvision model?

File: src/otx/algo/classification/vit.py [hunk 5]
```diff
@@ -494,10 +494,11 @@ def _build_model(self, head_config: dict) -> nn.Module:
...
             backbone=vit_backbone,
             neck=None,
-            head=HierarchicalLinearClsHead(
+            head=HierarchicalCBAMClsHead(
                 in_channels=vit_backbone.embed_dim,
                 multiclass_loss=nn.CrossEntropyLoss(),
                 multilabel_loss=AsymmetricAngularLossWithIgnore(gamma_pos=0.0, gamma_neg=1.0, reduction="sum"),
+                step_size=1,
                 **head_config,
             ),
```
Comment (harimkang) [inline src/otx/algo/classification/vit.py:503]: Does this also work well without adding `optimize_gap` to `False`?

**Review by harimkang — COMMENTED** [review]
LGTM, I left a few minor comments.

File: src/otx/algo/classification/heads/hlabel_cls_head.py (hunk 4 above)
Comment (sooahleex) [reply src/otx/algo/classification/heads/hlabel_cls_head.py]: Fix in [ed48f15](https://github.com/openvinotoolkit/training_extensions/pull/3810/commits/ed48f15864c74a8b14b3b5d7a68672d3e16f6ced)

File: src/otx/algo/classification/vit.py (hunk 5 above)
Comment (sooahleex) [reply src/otx/algo/classification/vit.py:503]: Fix in [ed48f15](https://github.com/openvinotoolkit/training_extensions/pull/3810/commits/ed48f15864c74a8b14b3b5d7a68672d3e16f6ced)

File: src/otx/algo/classification/torchvision_model.py
Comment (sooahleex) [reply src/otx/algo/classification/torchvision_model.py:1]: Fix in [ed48f15](https://github.com/openvinotoolkit/training_extensions/pull/3810/commits/ed48f15864c74a8b14b3b5d7a68672d3e16f6ced)

File: src/otx/algo/classification/heads/hlabel_cls_head.py (hunk 4 above)
Comment (sungchul1) [reply src/otx/algo/classification/heads/hlabel_cls_head.py]: Other parameters seem not to be updated, could you check it again?

**Review by harimkang — DISMISSED** [review]

**Review by harimkang — APPROVED** [review]

File: src/otx/algo/classification/heads/hlabel_cls_head.py (hunk 4 above)
Comment (sooahleex) [reply src/otx/algo/classification/heads/hlabel_cls_head.py]: I updated this in [e524909](https://github.com/openvinotoolkit/training_extensions/pull/3810/commits/e5249090e2a60624a32bca558ce177f41c4ce05e). Thank you for letting me know this!

File: src/otx/algo/classification/classifier/base_classifier.py [hunk 6]
```diff
@@ -61,6 +61,7 @@ def __init__(
         neck: nn.Module | None,
         head: nn.Module,
         pretrained: str | None = None,
+        optimize_gap: bool = True,
```
Comment (eunwoosh) [inline src/otx/algo/classification/classifier/base_classifier.py:64]: Could you add docstring for this argument?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work. I left a minor comment. Please take a look.

File: src/otx/algo/classification/classifier/base_classifier.py (hunk 6 above)
Comment (harimkang) [reply src/otx/algo/classification/classifier/base_classifier.py:64]: I'll update this in another PR https://github.com/openvinotoolkit/training_extensions/pull/3830 to refactor that. Let's leave it as it is.

**Review by eunwoosh — APPROVED** [review]

**Review by sungchul1 — APPROVED** [review]
